src/fsharp/LexFilter.fs

```diff
@@ -399,39 +399,62 @@ type LexbufState(startPos: Position,
     member x.EndPos = endPos
     member x.PastEOF = pastEOF
 
-[<Struct>]
-type PositionTuple =
-    val X: Position
-    val Y: Position
-    new (x: Position, y: Position) = { X = x; Y = y }
-
 /// Used to save the state related to a token
 [<Class>]
 type TokenTup = 
-    val Token : token
-    val LexbufState : LexbufState
-    val LastTokenPos: PositionTuple
+    // This is mutable for performance reasons.
+    val mutable Token : token
+    val mutable LexbufState : LexbufState
+    val mutable LastTokenPos: Position
     new (token, state, lastTokenPos) = { Token=token; LexbufState=state;LastTokenPos=lastTokenPos }
     
     /// Returns starting position of the token
     member x.StartPos = x.LexbufState.StartPos
     /// Returns end position of the token
     member x.EndPos = x.LexbufState.EndPos
-    
+        
+type TokenTupPool() =
+
+    [<Literal>]
+    let maxSize = 100
+
+    let stack = System.Collections.Generic.Stack(Array.init maxSize (fun _ -> TokenTup(Unchecked.defaultof<_>, Unchecked.defaultof<_>, Unchecked.defaultof<_>)))
+
+    member _.Rent() = 
+        if stack.Count = 0 then
+            assert false
+            TokenTup(Unchecked.defaultof<_>, Unchecked.defaultof<_>, Unchecked.defaultof<_>)
+        else
+            stack.Pop()
+
+    member _.Return(x: TokenTup) =
+        x.Token <- Unchecked.defaultof<_>
+        x.LexbufState <- Unchecked.defaultof<_>
+        x.LastTokenPos <- Unchecked.defaultof<_>
+        if stack.Count >= maxSize then
+            assert false
```

Comment:
If I understand correctly, throwing may become blocking issues to end users if a bug somehow slips in, while the assert will only affect people running debug compiler and hitting that bug.

I'm rephrasing the question with a bias toward keeping assert: Should a bug in the TokenTup lifecycle become a blocking issue for end user?

It feels to me that assert is safer but it will silent issues that may happen in real world usage, I'm interested to know the answer to your question :slightly_smiling_face: .

Reply:
What @smoothdeveloper said. While I haven't tested anything that relied on more than 100 `TokenTup` objects at once (I think at most it was ~30 when parsing TypeChecker.fs), it's a reasonable fallback just in case. This may silent the issue, but it's not necessarily a bug, it would just mean we would need to up the pool size. I believe in 99.9% of cases we will not see it ever reaching the maxsize.

Now, if a bug does occur in lifecycle, returning an object to the object pool clears out its properties and if anything still holds onto it, it will probably break something. However, if it passes all of CI, I feel confident that those cases have been handled.